Support the Wireshark plugin for more Wireshark distributions

BewareMyPower opened the pull request:
### Motivation

https://github.com/apache/pulsar-client-cpp/pull/182 fixes the Wireshark build process on macOS. However, it breaks the compatibility with some Wireshark distributions like the default Wireshark 3.2.3 on Ubuntu 20.04. The reason is some Wireshark distributions use `config.h` to record the versions, while other might use `ws_version.h`.

See
https://listman.redhat.com/archives/libvir-list/2020-September/msg00377.html for a similar fix.

### Modifications

Try to find the `ws_version.h` first, if it's not found, find the `config.h`.

To be more user friendly:
- Separate the dissector from the whole project so that we can build the dissector without building the Pulsar C++ client.
- Refactor the README of the Wireshark dissector by focusing on how to use.

### Verifications

Add a workflow to build Wireshark dissector on both Ubuntu and macOS.

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)

- [ ] `doc-not-needed` 
(Please explain why)

- [x] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

BewareMyPower commented:
@Demogorgon314 Could you help verify this CMakeLists.txt work on macOS? Not sure if `-Wl,-all_load` could be removed.

Demogorgon314 commented:
I tried it on macOS, and it has a compile error. 

```bash
➜  wireshark git:(bewaremypower/wireshark-fix) ✗ cmake -B build -DWIRESHARK_INCLUDE_PATH=/opt/homebrew/Cellar/wireshark/4.0.3/include
-- Use WIRESHARK_INCLUDE_PATH: /opt/homebrew/Cellar/wireshark/4.0.3/include
-- Use GLIB_INCLUDE_DIRS: /opt/homebrew/Cellar/glib/2.74.5/include/glib-2.0;/opt/homebrew/Cellar/glib/2.74.5/lib/glib-2.0/include;/opt/homebrew/opt/gettext/include;/opt/homebrew/Cellar/pcre2/10.42/include
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/wangkai/Developer/github/pulsar-client-cpp/wireshark/build

➜  wireshark git:(bewaremypower/wireshark-fix) ✗ cmake --build build                                                                 
[ 25%] Building CXX object CMakeFiles/pulsar-dissector.dir/pulsarDissector.cc.o
/Users/wangkai/Developer/github/pulsar-client-cpp/wireshark/pulsarDissector.cc:21:1: error: unknown type name 'constexpr'
constexpr int kWiresharkMajorVersion = WIRESHARK_VERSION_MAJOR;
^
/Users/wangkai/Developer/github/pulsar-client-cpp/wireshark/pulsarDissector.cc:22:1: error: unknown type name 'constexpr'
constexpr int kWiresharkMinorVersion = WIRESHARK_VERSION_MINOR;
^
In file included from /Users/wangkai/Developer/github/pulsar-client-cpp/wireshark/pulsarDissector.cc:37:
/Users/wangkai/Developer/github/pulsar-client-cpp/wireshark/build/generated/lib/PulsarApi.pb.h:10:10: fatal error: 'google/protobuf/port_def.inc' file not found
#include <google/protobuf/port_def.inc>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
3 errors generated.
gmake[2]: *** [CMakeFiles/pulsar-dissector.dir/build.make:83: CMakeFiles/pulsar-dissector.dir/pulsarDissector.cc.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/pulsar-dissector.dir/all] Error 2
gmake: *** [Makefile:91: all] Error 2

```

And if I don't specify the `WIRESHARK_INCLUDE_PATH`, it cannot find the path automatically.
```
-- Cannot find ws_version.h, fallback to find config.h
CMake Error at CMakeLists.txt:31 (message):
  Failed to find WIRESHARK_INCLUDE_PATH


-- Configuring incomplete, errors occurred!

```

Anonymitaet commented on wireshark/README.md:
```diff
@@ -60,50 +46,25 @@ $ brew install wireshark
+### Copy to the plugin directory
 
-Copy the dissector in the appropriate location so that Wireshark can find it at startup.
+Run the Wireshark, then click the menu `Help - About Wireshark - Plugins`, you will find the plugin directory. For example, it's `/usr/lib/x86_64-linux-gnu/wireshark/plugins/3.2/epan` on Ubuntu 20.04. Then, copy the dissector into that directory:
```
```suggestion
1. Run the Wireshark and click **Help - About Wireshark - Plugins**, and then you can find the plugin directory. For example, it's `/usr/lib/x86_64-linux-gnu/wireshark/plugins/3.2/epan` on Ubuntu 20.04. 

2. Copy the dissector into that directory.

```
In technical writing, we usually use bold for menus. Details see https://docs.google.com/document/d/1lc5j4RtuLIzlEYCBo97AC8-U_3Erzs_lxpkDuseU0n4/edit?pli=1#bookmark=kix.m5ccxnsf33mu

Anonymitaet commented on wireshark/README.md:
````diff
@@ -60,50 +46,25 @@ $ brew install wireshark
 ```
 
-### Complete installation
+To verify it has been loaded successfully, restart the Wireshark, you will see the plugin in the plugin list:
````
```suggestion
To verify whether it has been loaded successfully, restart the Wireshark and then you can see the plugin in the plugin list:
```

BewareMyPower commented:
@Anonymitaet The comments have been addressed now.

@Demogorgon314 I've fixed the build failure and added the workflow to build the Wireshark plugin on macOS, could you also verify whether it could be loaded?

Demogorgon314 commented:
@BewareMyPower Now it can build success and load, but it still needs manually specific `WIRESHARK_INCLUDE_PATH`.

BewareMyPower commented:
@Demogorgon314 It's confirmed that the install path `/opt/homebrew/Cellar/wireshark/4.0.3/include` on macOS M1 cannot be found by CMake directly. I've added back the document for users to specify the installation path of Wireshark.

Anonymitaet commented:
Thanks! Now the docs LGTM.

BewareMyPower commented on wireshark/CMakeLists.txt:
```diff
@@ -17,67 +17,46 @@
 
-MESSAGE(STATUS "Use WIRESHARK_BUILD_TYPE: ${CMAKE_BUILD_TYPE}")
+set(CMAKE_CXX_STANDARD 11)
 
```
Addressed.

Demogorgon314 replied:
The command should move in this line. Otherwise, it will use `#include <config.h>` by default, so when using a high version of Wireshark dependency will fail when built. 

```
➜  wireshark git:(bewaremypower/wireshark-fix) ✗ cmake --build build                         
[ 25%] Running cpp protocol buffer compiler on /Users/wangkai/Developer/github/pulsar-client-cpp/wireshark/../proto/PulsarApi.proto
[ 50%] Building CXX object CMakeFiles/pulsar-dissector.dir/pulsarDissector.cc.o
/Users/wangkai/Developer/github/pulsar-client-cpp/wireshark/pulsarDissector.cc:24:10: fatal error: 'config.h' file not found
#include <config.h>
         ^~~~~~~~~~
1 error generated.
gmake[2]: *** [CMakeFiles/pulsar-dissector.dir/build.make:84: CMakeFiles/pulsar-dissector.dir/pulsarDissector.cc.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/pulsar-dissector.dir/all] Error 2
gmake: *** [Makefile:91: all] Error 2

```

BewareMyPower replied:
I made this variable an extra search path in `find_path` (see https://cmake.org/cmake/help/latest/command/find_path.html). Does it work for you now?

Demogorgon314 replied:
Yes, now it works.
